core/src/collections/DynamicValue.hpp

```diff
@@ -28,47 +28,130 @@
+            explicit DynamicValue(int32_t x);
+            explicit DynamicValue(int64_t x);
+            explicit DynamicValue(double x);
+            explicit DynamicValue(const std::shared_ptr<DynamicArray>& x);
+            explicit DynamicValue(const std::shared_ptr<DynamicObject>& x);
 
             template <class Archive>
             void serialize(Archive& ar) {
-
             }
 
             std::string dump() const;
             std::ostream& dump(std::ostream& ss, int depth) const;
-        };
 
-        template<> void DynamicValue::serialize<cereal::PortableBinaryOutputArchive>( cereal::PortableBinaryOutputArchive & );
-        template<> void DynamicValue::serialize<cereal::PortableBinaryInputArchive>( cereal::PortableBinaryInputArchive & );
-        template<> void DynamicValue::serialize<cereal::BinaryOutputArchive>( cereal::BinaryOutputArchive & );
-        template<> void DynamicValue::serialize<cereal::BinaryInputArchive>( cereal::BinaryInputArchive & );
+            /// Get the underlying type.
+            api::DynamicType getType() const;
+
+            /// Change the value to a std::string by copying.
+            DynamicValue& operator=(const std::string& rhs);
+            /// Change the value to a std::string by moving in.
+            DynamicValue& operator=(std::string&& rhs);
+
+            /// Change the value to a std::vector<uint8_t> by copying.
+            DynamicValue& operator=(const std::vector<uint8_t>& rhs);
+            /// Change the value to a std::vector<uint8_t> by moving in.
+            DynamicValue& operator=(std::vector<uint8_t>&& rhs);
+
+            /// Change the value to a bool by copying.
+            DynamicValue& operator=(bool rhs);
+
+            /// Change the value to a int32_t by copying.
+            DynamicValue& operator=(int32_t rhs);
+
+            /// Change the value to a int64_t by copying.
+            DynamicValue& operator=(int64_t rhs);
+
+            /// Change the value to a double by copying.
+            DynamicValue& operator=(double rhs);
+
+            /// Change the value to a DynamicArray by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicArray>& rhs);
 
+            /// Change the value to a DynamicObject by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicObject>& rhs);
+
+            /// Try to get the value as if it were a std::string.
+            optional<std::string&> asStr();
+            /// Try to get the value as if it were a std::string.
+            optional<const std::string&> asStr() const;
+
+            /// Try to get the value as if it were a std::vector<uint8_t>.
+            optional<std::vector<uint8_t>&> asData();
+            /// Try to get the value as if it were a std::vector<uint8_t>.
+            optional<const std::vector<uint8_t>&> asData() const;
+
+            /// Try to get the value as if it were a bool.
+            optional<bool&> asBool();
```

Comment:
Why do we need to have a reference here? Why not optional<bool>?

Reply:
Semantic of this operation is not clear for me

Reply:
The former mutates `x`, the second one will create an error (because you will get `NONE`). If you want to change the type, you have to use `operator=`. And accessing the value of of an `optional<_>` should be done, in C++, by testing whether the value is there prior. So in your second example, you will get `hasValue() == false`.

Reply:
I found this dangerous.
```
auto x = DynamicValue(false);
auto y = x.asBool();
return y;
....
y.value = true; // this change x, that may not exist anymore
```

Reply:
This is typical C++ when someone tries to escape a reference from a local scope. It has nothing to do with the `optional` :) And I hope `msvc` / `clang` / `g++` will be able to see such an escaped reference!

Reply:
What is the benefit to have & inside optional? We can mutate DynamicValue directly if we want, & is not free, it still consume the same memory as pointer (so for bool it is even worse).
Taking this in mind and the danger nature of having optional<T&> I would go for more safe way - optional<T>. 
Looks like compilers don't want to help us here - https://ideone.com/4Pt2YH - runtime error.

Reply:
`optional<T&>` gives you “maybe” a reference that you can use to manipulate in a type-safe way the content of the `DynamicValue`. If it doesn’t contain a `T`, you don’t get a reference. Of course we’re talking about C++, so it’s easy to shoot oneself in the feet if you try to use the reference after having changed the `DynamicValue` or if you use it after free — Rust has escape analysis at least.

If you want to “keep a mutable reference” of a dynamic value and change its content without having to continuously check the type at runtime, this method gives you such semantics. It provides a better type-safe layer than accessing it via `int32` or that kind of things — look at the diff code, I removed a lot of dynamic type checking thanks to this function. However, I agree that it doesn’t solve the problem of escape analysis (we have the same problem with `std::vector<_>::data()`, for instance: you can get a pointer to something that has been freed, regrown or even moved to another memory place.

What you seem to be worried about here is the reference, not really the contract of those methods. I get your point but if we forbid people to do this, that means you cannot mutate the content of a `DynamicValue` without copying. If you think we don’t care, then I can remove the reference version, if you prefer, and return by copy.